week4/wordbreak.cc

```diff
@@ -0,0 +1,136 @@
+/* wordbreak.cc
+ * author: connie
+ *
+ * WORD BREAK
+ * Given an input string and a dictionary of words,
+ * find out if the input string can be segmented into a
+ * space-separated sequence of dictionary words.
+ *
+ */
+#include <iostream>
+#include <sstream>
+#include <cassert>
+#include <vector>
+#include <string>
+#include <map>
+
+using namespace std;
+
+//== DECLARATIONS
+bool wordInDict(string word, vector<string>& dict);
+bool wordBreak_RC(string s, vector<string>& dict);
+bool wordBreak_DP(string s, vector<string>& dict); // TODO w/ map
+
+//== DEFINITIONS
+//#define DEBUG
+
+//== FUNCTIONS
+/*
+ * bool wordInDict(string word, vector<string>& dict)
+ * inputs: word - word to check
+           dict - the dictionary (array of strings) with valid words
+ * outputs: True for valid dictionary word, False for invalid word
+ * notes: helper function for iterating thru the array
+ */
+bool wordInDict(string word, vector<string>& dict)
+{
+  if (word.size() == 0) return false;
+
+  // iterate thru dictionary to check if the word is valid
+  for (string dictword : dict) {
+#ifdef DEBUG
+    cout << word << " ?= " << dictword << endl;
+#endif
+    if (word.compare(dictword) == 0) return true;
+  }
+
+  return false;
+}
+
+/*
+ * bool wordBreak_RC(string s, vector<string>& dict)
+ * inputs: s - the string to check for a valid "word break"
+           dict - the dictionary (array of strings) with valid words
+ * outputs: True for valid word break, False for invalid word break
+ * notes: the recursive implementation that verifies progressive substrings
+ */
+bool wordBreak_RC(string s, vector<string>& dict)
+{
+  // base case: if nothing is remaining then we are done
+  if (s.size() == 0) return true;
+
+  // we want to iterate through the string from s(0,1) - the first substring
+  for (int i = 1; i <= s.size(); i++) {
+
+    // if substring that we are looking at is in dict, check if next part of
+    // the string is also in the dict (either as word or as compound word)
+    // by recursing on the remainder of the string
+    if (
+      wordInDict(s.substr(0,i),dict) && wordBreak_RC(s.substr(i,s.size()-i),dict)
+    ) return true;
+  }
+
+  // if we reach this point then there are still parts of string not in dict
+  return false;
+}
+
+/** main function for recursion **/
+bool wordBreak(string s, vector<string> dict)
+{
+  // handle empty cases
+  if (s.size() == 0 || dict.size() == 0) return false;
+
+  // word breaks aren't case sensitive
```

Comment:
Good job thinking to do this during set-up instead of having to do redundant conversions later on :)